File: brightcove-exoplayer/BasicCastSampleApp/src/main/java/com/brightcove/player/samples/cast/basic/VideoListAdapter.java
package com.brightcove.player.samples.cast.basic;

import android.annotation.SuppressLint;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.recyclerview.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import com.brightcove.player.model.Video;
import com.squareup.picasso.Picasso;

import java.net.URI;
import java.util.HashMap;
import java.util.List;
import java.util.Locale;
import java.util.Map;
import java.util.concurrent.TimeUnit;

import static com.brightcove.player.samples.cast.basic.VideoPlayerActivity.PROPS_SHORT_DESCRIPTION;

public class VideoListAdapter extends RecyclerView.Adapter<VideoListAdapter.ViewHolder> {

    /**
     * The current list of videos.
     */
    private List<Video> videoList;

    /**
     * A map of video identifiers and position of video in the list.
     */
    @SuppressWarnings("MismatchedQueryAndUpdateOfCollection")
    private Map<String, Integer> indexMap = new HashMap<>();

    private final ItemClickListener clickListener;
    public interface ItemClickListener {
        void itemClicked(View view, Video video, int position);
    }

    VideoListAdapter(ItemClickListener clickListener) {
        this.clickListener = clickListener;
    }

    @Override
    public ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(parent.getContext()).inflate(R.layout.video_item_view, parent, false);

        return new ViewHolder(view);
    }

    @Override
    public void onBindViewHolder(ViewHolder holder, @SuppressLint("RecyclerView") final int position) {
        final Video video = videoList.get(position);
        holder.videoTitleText.setText(video.getName());

        Object descriptionObj = video.getProperties().get(PROPS_SHORT_DESCRIPTION);
        if (descriptionObj instanceof String) {
            holder.videoDescriptionText.setText((String) descriptionObj);
        }

        int duration = video.getDuration();
        if (duration > 0) {
            holder.videoDurationText.setText(millisecondsToString(duration));
            holder.videoDurationText.setVisibility(View.VISIBLE);
        } else {
            holder.videoDurationText.setText(null);
            holder.videoDurationText.setVisibility(View.GONE);
        }

        URI imageUri = video.getStillImageUri();
        if (imageUri == null) {
            holder.videoThumbnailImage.setImageResource(R.drawable.movie);
        } else {
            Picasso.with(holder.itemView.getContext()).load(imageUri.toASCIIString()).into(holder.videoThumbnailImage);
        }

        holder.videoThumbnailImage.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                clickListener.itemClicked(view, video, position);
            }
        });
    }

    @Override
    public int getItemCount() {
        return videoList == null ? 0 : videoList.size();
    }

    class ViewHolder extends RecyclerView.ViewHolder {

        /**
         * Reference to the thumbnail image view.
         */
        final ImageView videoThumbnailImage;
        /**
         * Reference to the video title view.
         */
        final TextView videoTitleText;
        /**
         * Reference to the video status information text view.
         */
        final TextView videoDescriptionText;
        /**
         * Reference to the video duration view.
         */
        final TextView videoDurationText;

        ViewHolder(View itemView) {
            super(itemView);
            videoThumbnailImage = itemView.findViewById(R.id.video_thumbnail_image);
            videoTitleText = itemView.findViewById(R.id.video_title_text);
            videoDurationText = itemView.findViewById(R.id.video_duration_text);
            videoDescriptionText = itemView.findViewById(R.id.video_description_text);

        }
    }

    /**
     * Sets the list of videos to be shown. If this value is null, then the list will be empty.
     *
     * @param videoList list of {@link Video} objects.
     */
    void setVideoList(@Nullable List<Video> videoList) {
        this.videoList = videoList;
        buildIndexMap();
    }

    /**
     * Build the index map.
     */
    private void buildIndexMap() {
        indexMap.clear();
        if (videoList != null) {
            int index = 0;
            for (Video video : videoList) {
                indexMap.put(video.getId(), index++);
            }
        }
        notifyDataSetChanged();
    }

    /**
     * Converts the given duration into a time span string.
     *
     * @param duration elapsed time as number of milliseconds.
     * @return the formatted time span string.
     */
    @NonNull
    private static String millisecondsToString(long duration) {
        final TimeUnit scale = TimeUnit.MILLISECONDS;

        StringBuilder builder = new StringBuilder();
        long days = scale.toDays(duration);
        duration -= TimeUnit.DAYS.toMillis(days);
        if (days > 0) {
            builder.append(days);
            builder.append(days > 1 ? " days " : " day ");
        }

        long hours = scale.toHours(duration);
        duration -= TimeUnit.HOURS.toMillis(hours);
        if (hours > 0) {
            builder.append(String.format(Locale.getDefault(),"%02d:", hours));
        }

        long minutes = scale.toMinutes(duration);
        duration -= TimeUnit.MINUTES.toMillis(minutes);

        long seconds = scale.toSeconds(duration);
        builder.append(String.format(Locale.getDefault(),"%02d:%02d", minutes, seconds));

        return builder.toString();
    }
}

File: brightcove-exoplayer-kotlin/BasicCastGoogleReceiverSampleApp/src/main/java/com/brightcove/googlecastreceiver/VideoListAdapter.kt
package com.brightcove.googlecastreceiver

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.ListAdapter
import androidx.recyclerview.widget.RecyclerView
import com.brightcove.googlecastreceiver.databinding.ItemViewBinding
import com.brightcove.player.model.Video
import com.squareup.picasso.Picasso
import java.util.*
import java.util.concurrent.TimeUnit


typealias OnVideoClick = (View, Video) -> Unit

class VideoListAdapter(private val onVideoClick: OnVideoClick) : ListAdapter<Video, VideoListAdapter.ViewHolder>(DIFF_CALLBACK) {

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        val binding = ItemViewBinding.inflate(LayoutInflater.from(parent.context), parent, false)
        return ViewHolder(binding, onVideoClick)
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        //Get Video information
        holder.bind(getItem(position))
    }

    class ViewHolder (
        private val binding: ItemViewBinding,
        private val onVideoClick: OnVideoClick
    ) : RecyclerView.ViewHolder(binding.root) {

        fun bind(video: Video){

            binding.apply {

                //Set Video info
                videoTitleText.text = video.name

                val descriptionObj: String? = video.properties[VideoPlayerActivity.PROPS_SHORT_DESCRIPTION]?.toString()

                if(!descriptionObj.isNullOrBlank())
                    videoDurationText.text = descriptionObj

                if (video.durationLong > 0) {
                    videoDurationText.text = millisecondsToString(video.durationLong)
                    videoDurationText.visibility = View.VISIBLE
                } else {
                    videoDurationText.text = ""
                    videoDurationText.visibility = View.GONE
                }

                val imageUri = video.stillImageUri
                if (imageUri == null) {
                    videoThumbnailImage.setImageResource(R.drawable.movie)
                } else {
                    Picasso.get()
                        .load(imageUri.toASCIIString())
                        .into(videoThumbnailImage)
                }

                //Adding onCLick Listener to the view
                videoThumbnailImage.setOnClickListener{
                    onVideoClick(this.root,video)
                }
            }
        }

        /**
         * Converts the given duration into a time span string.
         *
         * @param durationLong elapsed time as number of milliseconds.
         * @return the formatted time span string.
         */
        private fun millisecondsToString(durationLong: Long): String {
            var duration = durationLong
            val scale = TimeUnit.MILLISECONDS
            val builder = StringBuilder()
            val days = scale.toDays(duration)
            duration -= TimeUnit.DAYS.toMillis(days)
            if (days > 0) {
                builder.append(days)
                builder.append(if (days > 1) " days " else " day ")
            }
            val hours = scale.toHours(duration)
            duration -= TimeUnit.HOURS.toMillis(hours)
            if (hours > 0) {
                builder.append(String.format(Locale.getDefault(), "%02d:", hours))
            }
            val minutes = scale.toMinutes(duration)
            duration -= TimeUnit.MINUTES.toMillis(minutes)
            val seconds = scale.toSeconds(duration)
            builder.append(String.format(Locale.getDefault(), "%02d:%02d", minutes, seconds))
            return builder.toString()
        }
    }

    companion object {
        val DIFF_CALLBACK = object : DiffUtil.ItemCallback<Video>() {
            override fun areItemsTheSame(oldItem: Video, newItem: Video): Boolean =
                oldItem.id == newItem.id

            override fun areContentsTheSame(oldItem: Video, newItem: Video): Boolean =
                oldItem.equals(newItem)
        }
    }
}